concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+
+Cardinality
+-----------
+
+"Cardinality" means "[the number of elements in a set](https://en.wikipedia.org/wiki/Cardinality)".
+
+In IPLD, we use the phrase "the cardinality of a type" to define how many members a type can have.
+
+### Why is Cardinality important?
+
+Cardinality is important to the way we define and think about our Data Model and Schema type systems,
+because if the cardinality of two parts of a model aren't the same cardinality,
+then that means one of them is less expressive.
+If we intended two systems to be interchangeable, but they have unequal cardinality,
+then we can quickly and clearly identify that there must be a bug somewhere!
+
+Cardinality concepts are also useful for clear documentation.
+
+Cardinality counting can be used as a design planning exercise:
+count how many states your types *can* have, and then consider how many states
+your code is designed to handle, and if they're not equal,
+then it's clear that you have work to do!
+
+In practical application, you can see some examples of cardinality-counting used
+to describe `nullable` vs `optional` vs `implicit` values in the IPLD Schemas documentation:
+[Cardinality Examples](../schemas/schema-kinds.md#cardinality-examples).
+These cardinality countings are both key to the documentation,
+and key to understanding and using those features to build unambiguous APIs.
+
+### How do we Count Cardinality?
+
+Counting cardinality is probably best done by example.
+
+Consider a boolean: it can have the value 'true', or it can have the value 'false'.
+We simply count these states up!  The cardinality of a boolean is `2`!
+
+Now, consider a pair of boolean values: how many states can it have?
+Four: `(true, true)`, `(true, false)`, `(false, true)`, or `(false, false)`.
+So, the cardinality of two booleans is `4`.
+
+The exercise gets more interesting as we consider larger accumulations of values.
+
+Consider three boolean values: how many states can it have?
+(You can draw the full table yourself, for this one.)
+When you count them all up, you should reach `8`.
+
+But now, let's imagine something new: suppose we've got some sort of values
+that we call "nooleans".  Nooleans can have the value 'nue' or 'fuu'.
+
+We'll explore more examples of this, and start identifying categories of
+things where we can do predictable kinds of math with cardinality in
+the next section: [Categories of Types](#categories-of-types).
+(Spoiler: the examples above correspond to an enum, then a product type,
+then another bigger product type, and finally a sum type.)
+
+
+Categories of Types
+-------------------
+
+We can identify a few categories of "types" based on how we evaluate their cardinality.
+(These definitions will use names that are popularized in computer science,
+and so they *may* match the name used in your favorite programming language -- but
+whether or not the same is familiar, the *concept* is universal, even across languages!)
+
+### enums
```

Comment:
I would suggest using "enumeration" here rather than the language-keyword abbreviation, which, as you've noted, is prone to causing an incorrect understanding of the underlying concept. (Well, I guess that's only true of Rust. C enums are bad for many reasons, but "naming accuracy" is not one of them.)